File: RepositoryPattern/SampleApplicationForGM/app/src/main/java/com/example/sampleapplicationforgm/viewmodel/MainActivityViewModel.java
package com.example.sampleapplicationforgm.viewmodel;

import com.example.basicapp.event.LaunchActivityEvent;
import com.example.basicapp.ui.DetailActivity;
import com.example.myfeature.MyFeatureService;
import com.example.myfeature.models.DummyResponse;

import java.util.List;

import javax.inject.Inject;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.lifecycle.ViewModel;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class MainActivityViewModel extends ViewModel implements LifecycleObserver {

    private MyFeatureService myFeatureService;
    public MutableLiveData<Boolean> shouldShowDescription = new MutableLiveData<>();
    public MutableLiveData<LaunchActivityEvent> launchDetailActivity = new MutableLiveData<>();

    @Inject
    public MainActivityViewModel(MyFeatureService myFeatureService) {
        this.myFeatureService = myFeatureService;
    }

    public void doApiCall() {
        shouldShowDescription.setValue(true);
        Disposable disposable = myFeatureService.getDummyRestResponse()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> { System.out.println("***Success Response: ");
                    updateResponse(response);} , throwable -> System.out.println("***Error: " + throwable.getMessage()), () -> System.out.println("On Complete"));
    }

    private void updateResponse(List<DummyResponse> response) {
        for(DummyResponse dummyResponse : response) {
            System.out.println("***Employee ID:  " + dummyResponse.getId());
            System.out.println("***Employee Name:  " + dummyResponse.getName());
            System.out.println("***Employee Salary:  " + dummyResponse.getSalary());
        }
        shouldShowDescription.setValue(false);
    }

    public void launchDetailScreen() {
        launchDetailActivity.setValue(new LaunchActivityEvent(DetailActivity.class));
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onCreate() {
        System.out.println("MainActivityViewModel onCreate called ");
        shouldShowDescription.setValue(false);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    public void onResume() {
        System.out.println("MainActivityViewModel onResume called ");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    public void onStart() {
        System.out.println("MainActivityViewModel onStart called ");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    public void onStop() {
        System.out.println("MainActivityViewModel onStop called ");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    public void onPause() {
        System.out.println("MainActivityViewModel onPause called ");
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroy() {
        System.out.println("MainActivityViewModel onDestroy called ");
    }

    @Override
    protected void onCleared() {
        super.onCleared();
    }
}

File: RepositoryPattern/SampleApplicationForGM/app/src/main/java/com/example/sampleapplicationforgm/viewmodel/MainActivityViewModel.kt
package com.example.sampleapplicationforgm.viewmodel

import androidx.lifecycle.*
import com.example.myfeature.MyFeatureService
import com.example.myfeature.models.DummyResponse
import com.example.sampleapplicationforgm.event.LaunchActivityEvent
import com.example.sampleapplicationforgm.ui.DetailActivity
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

class MainActivityViewModel @Inject constructor(private val myFeatureService: MyFeatureService) : ViewModel(), LifecycleObserver {
    @JvmField
    var shouldShowDescription = MutableLiveData<Boolean>()
    var launchDetailActivity = MutableLiveData<LaunchActivityEvent>()
    fun doApiCall() {
        shouldShowDescription.value = true
        val disposable = myFeatureService.dummyRestResponse
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response: List<DummyResponse> ->
                    println("***Success Response: ")
                    updateResponse(response)
                }, { throwable: Throwable -> println("***Error: " + throwable.message) }) { println("On Complete") }
    }

    private fun updateResponse(response: List<DummyResponse>) {
        for (dummyResponse in response) {
            println("***Employee ID:  " + dummyResponse.id)
            println("***Employee Name:  " + dummyResponse.name)
            println("***Employee Salary:  " + dummyResponse.salary)
        }
        shouldShowDescription.value = false
    }

    fun launchDetailScreen() {
        launchDetailActivity.value = LaunchActivityEvent(DetailActivity::class.java)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onCreate() {
        println("MainActivityViewModel onCreate called ")
        shouldShowDescription.value = false
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        println("MainActivityViewModel onResume called ")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onStart() {
        println("MainActivityViewModel onStart called ")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_STOP)
    fun onStop() {
        println("MainActivityViewModel onStop called ")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun onPause() {
        println("MainActivityViewModel onPause called ")
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        println("MainActivityViewModel onDestroy called ")
    }

    override fun onCleared() {
        super.onCleared()
    }

}